akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -96,208 +105,222 @@ private[http] object HttpServerBluePrint {
+      var oneHundredContinueResponsePending = false
+      var pullSuppressed = false
+      var messageEndPending = false
+
+      setHandler(requestParsingIn, new InHandler {
+        def onPush(): Unit =
+          grab(requestParsingIn) match {
             case r: RequestStart ⇒
-              requestStart = r
-              pull(applicationInput)
-              if (r.expect100ContinueResponsePending)
-                read(oneHundredContinue) { cont ⇒
-                  emit(out, ResponseRenderingContext(HttpResponse(StatusCodes.Continue)))
-                  requestStart = requestStart.copy(expect100ContinueResponsePending = false)
-                }
-            case MessageStartError(status, info) ⇒ finishWithError(status, info)
-          }
-        }
-        override def onUpstreamFinish(): Unit =
-          requestStart match {
-            case null ⇒ completeStage()
-            case r    ⇒ requestStart = r.copy(closeRequested = true)
+              openRequests = openRequests.enqueue(r)
+              messageEndPending = r.createEntity.isInstanceOf[StreamedEntityCreator[_, _]]
+              val rs = if (r.expect100Continue) {
+                oneHundredContinueResponsePending = true
+                r.copy(createEntity = with100ContinueTrigger(r.createEntity))
+              } else r
+              push(requestPrepOut, rs)
+            case MessageEnd ⇒
+              messageEndPending = false
+              push(requestPrepOut, MessageEnd)
+            case MessageStartError(status, info) ⇒ finishWithIllegalRequestError(status, info)
+            case x                               ⇒ push(requestPrepOut, x)
           }
+        override def onUpstreamFinish() =
+          if (openRequests.isEmpty) completeStage()
+          else complete(requestPrepOut)
       })
 
-      setHandler(applicationInput, new InHandler {
-        override def onPush(): Unit = {
-          val response = grab(applicationInput)
-          // see the comment on [[OneHundredContinue]] for an explanation of the closing logic here (and more)
-          val close = requestStart.closeRequested || requestStart.expect100ContinueResponsePending
-          abortReading(oneHundredContinue)
-          emit(out, ResponseRenderingContext(response, requestStart.method, requestStart.protocol, close),
-            if (close) () ⇒ completeStage() else pullBypass)
+      setHandler(requestPrepOut, new OutHandler {
+        def onPull(): Unit =
+          if (oneHundredContinueResponsePending) pullSuppressed = true
+          else pull(requestParsingIn)
+        override def onDownstreamFinish() = cancel(requestParsingIn)
+      })
+
+      setHandler(httpResponseIn, new InHandler {
+        def onPush(): Unit = {
+          val response = grab(httpResponseIn)
+          val requestStart = openRequests.head
+          openRequests = openRequests.tail
+          val isEarlyResponse = messageEndPending && openRequests.isEmpty
```

Comment:
this means that in the face of pipelining early responses are not detected; fixing this would be highly welcome :-) (since we’ll want to piggy-back on the code in order to consume non-materialized request entities at this point as well)